### PR TITLE
Keypoint loss refactor

### PR DESCRIPTION
**Pull request by kozlov721**
Loss and head refactor

### CODE REVIEW

File: src/luxonis_train/models/heads/keypoint_box_head.py [hunk 1]
```diff
@@ -81,7 +81,8 @@ def __init__(
         self.n_kpt_out = 3 * self.n_keypoints
         self.n_out = self.n_det_out + self.n_kpt_out
         self.n_anchors = len(anchors[0]) // 2
-        self.grid = [torch.zeros(1)] * self.num_heads
+        self.box_offset = 5
```
Comment (tersekmatija) [inline src/luxonis_train/models/heads/keypoint_box_head.py:84]: Would make sense defining this higher and using it for example for `self.n_det_out`.

File: src/luxonis_train/models/heads/keypoint_box_head.py [hunk 2]
```diff
@@ -135,63 +137,160 @@ def forward(self, inputs):
...
+        out_bbox = bbox.sigmoid()
+        out_bbox_xy = (out_bbox[..., 0:2] * 2.0 - 0.5 + grid) * stride
+        out_bbox_wh = (out_bbox[..., 2:4] * 2) ** 2 * anchor_grid.view(
+            1, self.n_anchors, 1, 1, 2
+        )
+        return torch.cat((out_bbox_xy, out_bbox_wh, out_bbox[..., 4:]), dim=-1)
+
+    def postprocess_for_loss(
+        self, output: Tuple[Tensor, List[Tensor]], label_dict: Dict[str, Tensor]
+    ) -> Tuple[
```
Comment (tersekmatija) [inline src/luxonis_train/models/heads/keypoint_box_head.py:201]: Let's document the output.

File: src/luxonis_train/utils/losses/box_loss.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
+from typing import Tuple
+
+import torch
+from torch import Tensor, nn
+
+from luxonis_train.utils.boxutils import bbox_iou
+
+
+class BoxLoss(nn.Module):
```
Comment (tersekmatija) [inline src/luxonis_train/utils/losses/box_loss.py:9]: Would it make sense moving _infer_box from the head to some util file and then using _infer_box inside this loss function?

File: src/luxonis_train/utils/losses/keypoint_loss.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
+from typing import Tuple
+
+import torch
+from torch import Tensor, nn
+
+from .common import BCEWithLogitsLoss
+
+class KeypointLoss(nn.Module):
```
Comment (tersekmatija) [inline src/luxonis_train/utils/losses/keypoint_loss.py:8]: Same comment here for _infer_keypoints function.

File: src/luxonis_train/utils/losses/keypoint_loss.py [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+        for a given prediction and target.
+
+        Args:
+            prediction (Tensor): The predicted tensor of shape
+                (n_detections, n_keypoints * 3).
+            target (Tensor): The target tensor of shape (n_detections, n_keypoints * 2).
+
+        Returns:
+            Tuple[Tensor, Tensor]: A tuple containing the keypoint loss
+            tensor of shape (1,) and the visibility loss tensor of shape (1,).
```
Comment (tersekmatija) [inline src/luxonis_train/utils/losses/keypoint_loss.py:25]: Should we link the paper with the equation for the keypoints loss?

**Review by tersekmatija — COMMENTED** [review]
Thanks, LGTM, left some minor comments.

**Review by klemen1999 — APPROVED** [review]
LGTM, when merged we'll probably want to make some small changes regarding BaseLoss implemented in this [PR](https://github.com/luxonis/models/pull/23) (won't have any effect to actual logic)
